python/sglang/srt/layers/linear.py

```diff
@@ -0,0 +1,509 @@
+import logging
```

Comment:
I'm a bit confused by the purpose of this file. My understanding is that you want to have the "actual tp size" applied for KV Linear Proj while tp size for Q Proj and O Proj. However, the `get_num_kv_head` already handled this case by default( see the max operation in https://github.com/vllm-project/vllm/blob/3eeb148f467e3619e8890b1a5ebe86a173f91bc9/vllm/config.py#L372-L380)

Reply:
I think it is the GQA that makes this part complicated. Because now we applied different parallelisms for Q/O proj and KV proj, we need to manually handle the mapping between Q heads and K/V heads. 

For example, suppose that we have TP1 x SP2, 8 Q heads `[q0, q1, ..., q7]`, and 2 KV heads `[k0, k1]`. GQA assumes two groups, where `[q0, q1, q2, q3] -> [k0]`, and `[q4, q5, q6, q7] -> [k1]`. Then for SP worker 0, it should be assigned Q heads `[q0, q1, q4, q5]`, and KV heads `[k0, k1]`. So this file adapts original vllm linear layers but changes the weight loader to select the correct set of Q/O heads (https://github.com/ivanium/sglang/blob/seq-parallel-kernel/python/sglang/srt/layers/linear.py#L486)

Reply:
We don't have the case right? The TP is always >= SP

Reply:
My bad. I meant TP2 x SP2

Reply:
now kind of understand your point